rp2040-hal/src/xosc.rs

```diff
@@ -0,0 +1,192 @@
+pub struct Initialized {
+    freq_hz: Hertz<u64>
+}
+
+/// Stable state (typestate)
+pub struct Stable{
+    freq_hz: Hertz<u64>
+}
+
+/// XOSC is disabling (typestate)
+pub struct Disabling;
+
+/// XOSC is in dormant mode (see Chapter 2, Section 16, §5)
+pub struct Dormant;
+
+impl State for Disabled {}
+impl State for Initialized {}
+impl State for Stable {}
+impl State for Disabling {}
+impl State for Dormant {}
+
+/// Possible errors when initializing the CrystalOscillator
+pub enum Error {
+    /// Frequency is out of the 1-15MHz range (see datasheet)
+    FrequencyOutOfRange,
+
+    /// Argument is bad : overflows, ...
+    BadArgument
+}
+
+/// A Crystal Oscillator.
+pub struct CrystalOscillator<S: State> {
+    device: rp2040_pac::XOSC,
+    state: S
+}
+
+impl<S: State> CrystalOscillator<S> {
+    /// Transitions the oscillator to another state.
+    fn transition<To: State>(self, state: To) -> CrystalOscillator<To> {
+        CrystalOscillator {
+            device: self.device,
+            state: state
+        }
+    }
+
+    /// Releases the underlying device.
+    pub fn free(self) -> rp2040_pac::XOSC {
+        self.device
+    }
+}
+
+impl CrystalOscillator<Disabled> {
+
+    /// Creates a new CrystalOscillator from the underlying device.
+    pub fn new(dev: rp2040_pac::XOSC) -> Self {
+        CrystalOscillator {
+            device: dev,
+            state: Disabled
+        }
+    }
+
+    /// Initializes the XOSC : frequency range is set, startup delay is calculated and set.
+    pub fn initialize(self, frequency: Megahertz<u32>) -> Result<CrystalOscillator<Initialized>, Error> {
+
+        const ALLOWED_FREQUENCY_RANGE: RangeInclusive<Megahertz<u32>> = Megahertz(1)..=Megahertz(15);
+        const STABLE_DELAY: Milliseconds<u64> = Milliseconds(1_u64);
+        const DIVIDER: Fraction = Fraction::new(1, 256);
+
+        if !ALLOWED_FREQUENCY_RANGE.contains(&frequency) {
+            return Err(Error::FrequencyOutOfRange)
+        }
+
+        self.device.ctrl.write(|w| {
+            w.freq_range()._1_15mhz();
+            w
+        });
+
+        let freq_hz: Hertz<u64> = frequency.into();
+        let delay_sec: Seconds<u64> = STABLE_DELAY.into();
+
+        //let startup_delay = ((freq_hz / 1000) + 128) / 256;
+        let startup_delay = delay_sec.
+            checked_mul(freq_hz.integer()).and_then(|r|
+                 r.to_generic::<u64>(DIVIDER).ok()
+             ).
+            ok_or(Error::BadArgument)?;
+
+        self.device.startup.write(|w| unsafe {
+            w.delay().bits(*startup_delay.integer() as u16);
+            w
+        });
+
+        self.device.ctrl.write(|w| {
+            w.enable().enable();
+            w
+        });
+
+        Ok(self.transition(Initialized {
+            freq_hz: freq_hz
+        }))
+    }
+}
+
+/// A token that's given when the oscillator is stablilzed, and can be exchanged to proceed to the next stage.
+pub struct StableOscillatorToken {
```

Comment:
Oh this is nice, did you see this in my previous implementation or is this a common pattern? :)

Reply:
At first I did without because I written a `_blocking` method with a `while { nop }` inside. Then, as I wanted to implement what I would call the "WouldBlock" pattern, I realized your use of these tokens was actually a very elegant solution !

Reply:
Thanks :) I was actually a bit confused because most of the code looked new but the token looked familiar.

In the end I added a blocking helper method (`new_stable_blocking()`) that runs through all the steps for cases where you just want a running Oscilator without caring about all the steps. Maybe this could be a good adition to this driver as well?

Reply:
And done :)